mapiproxy/servers/default/nspi/dcesrv_exchange_nsp.c

```diff
@@ -235,68 +235,75 @@ static void dcesrv_NspiUnbind(struct dcesrv_call_state *dce_call,
-able to be called from NspiQueryRows without doing again the same sanity checks and reusing the emsabp_ctx
-*/
-static void dcesrv_do_NspiUpdateStat(TALLOC_CTX *mem_ctx, struct NspiUpdateStat *r, struct emsabp_context *emsabp_ctx)
-{
-	enum MAPISTATUS			retval = MAPI_E_SUCCESS;
-	enum MAPISTATUS                 ret;
-	uint32_t			row, row_max;
-	TALLOC_CTX			*local_mem_ctx;
-	struct PropertyTagArray_r	*mids;
-
-	local_mem_ctx = talloc_zero(NULL, TALLOC_CTX);
 
-	mids = talloc_zero(local_mem_ctx, struct PropertyTagArray_r);
-        DCESRV_NSP_RETURN_IF(!mids, r, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
+static void position_in_table(struct STAT *pStat, struct emsabp_context *emsabp_ctx,
+			      struct PropertyTagArray_r *mids,
+			      uint32_t *out_row, uint32_t *out_last_row)
+{
+	bool		found;
+	uint32_t	row;
+	uint32_t	last_row;
 
-	ret = emsabp_search(local_mem_ctx, emsabp_ctx, mids, NULL, r->in.pStat, 0);
-	DCESRV_NSP_RETURN_IF(ret == MAPI_E_CALL_FAILED, r, ret, local_mem_ctx);
-	if (ret != MAPI_E_SUCCESS) {
-		row_max = 0;
+	if (mids->cValues > 0) {
+		last_row = mids->cValues - 1;
 	} else {
-		row_max = mids->cValues - 1;
+		last_row = 0;
 	}
 
-	if (r->in.pStat->CurrentRec == MID_CURRENT) {
+	if (pStat->CurrentRec == MID_CURRENT) {
 		/* Fractional positioning (3.1.4.5.2) */
-		row = r->in.pStat->NumPos * row_max / r->in.pStat->TotalRecs;
-		if (row > row_max) {
-			row = row_max;
+		row = pStat->NumPos * (last_row+1) / pStat->TotalRecs;
+		if (row > last_row) {
+			row = last_row;
 		}
 	} else {
 		/* Absolute positioning (3.1.4.5.1) */
-		if (r->in.pStat->CurrentRec == MID_BEGINNING_OF_TABLE) {
+		if (pStat->CurrentRec == MID_BEGINNING_OF_TABLE) {
 			row = 0;
 		}
-		else if (r->in.pStat->CurrentRec == MID_END_OF_TABLE) {
-			row = row_max;
+		else if (pStat->CurrentRec == MID_END_OF_TABLE) {
+			row = last_row;
 		} else {
-			retval = MAPI_E_NOT_FOUND;
+			found = false;
 			row = 0;
-			while (row <= row_max) {
-				if ((uint32_t) mids->aulPropTag[row] == (uint32_t) r->in.pStat->CurrentRec) {
-					retval = MAPI_E_SUCCESS;
+			while (row <= last_row) {
+				if (mids->aulPropTag[row] == pStat->CurrentRec) {
+					found = true;
 					break;
-				}
-				else {
+				} else {
 					row++;
 				}
 			}
-			if (retval == MAPI_E_NOT_FOUND) {
-				goto end;
+			if (!found) {
+				/* In this case the position is undefined. To avoid problems we will use first row */
+				row = 0;
 			}
 		}
 	}
 
+	*out_row = row;
+	*out_last_row = last_row;
+}
+
+
+/**
+This method does the work of NspiUpdateStat, it is separated from dcesrv_NspiUpdateStat to be
+able to be called from NspiQueryRows without doing again the same sanity checks and reusing the emsabp_ctx
+*/
+static void dcesrv_do_NspiUpdateStat(TALLOC_CTX *mem_ctx, struct NspiUpdateStat *r,
+				     struct emsabp_context *emsabp_ctx, struct PropertyTagArray_r *mids)
+{
+	enum MAPISTATUS			retval = MAPI_E_SUCCESS;
+	uint32_t			row, last_row;
+
```

Comment:
Sanity check on r->in.pStat is also required (assuming NspiUpdateState pulled payload can actually have a null pStat - which I have not checked).